File: src/lib/cloud/vault.ts
/**
 * HashiCorp Vault client (KV v2) for cloud credentials.
 *
 * The Vault CONNECTION (URL + token) is resolved per-user from the DB
 * (`VaultConfig`, see vault-config.ts), falling back to the VAULT_ADDR /
 * VAULT_TOKEN env vars. AWS access key + secret are stored in Vault (never in
 * Postgres) and the runner/agent read them back at execution time. Only a
 * non-secret *path* is kept in the database (`CloudProvider.credVaultPath`).
 *
 * KV v2 REST shape:
 *   write: POST  {addr}/v1/{mount}/data/{path}   body { data: {...} }
 *   read:  GET   {addr}/v1/{mount}/data/{path}  -> { data: { data: {...} } }
 *   del:   DELETE{addr}/v1/{mount}/metadata/{path}   (removes all versions)
 */
import {
  resolveVaultConn,
  resolveVaultConnForProvider,
  type VaultConn,
} from "./vault-config";

export type AwsKeys = {
  accessKeyId: string;
  secretAccessKey: string;
  region?: string;
};

const DEFAULT_PREFIX = () => process.env.VAULT_PATH_PREFIX?.trim() || "dda/cloud";

/**
 * Sync env-only check. Still used where we only care whether the platform-level
 * env Vault is set. For the per-user picture use `getVaultConfigView`.
 */
export function vaultConfigured(): boolean {
  return Boolean(process.env.VAULT_ADDR?.trim() && process.env.VAULT_TOKEN?.trim());
}

/** Deterministic, per-provider Vault path. Never exposes secrets in the DB. */
export function providerVaultPath(providerId: string, prefix?: string): string {
  return `${(prefix ?? DEFAULT_PREFIX()).replace(/\/+$/, "")}/${providerId}`;
}

async function vaultFetch(
  conn: VaultConn,
  method: "GET" | "POST" | "DELETE",
  path: string,
  body?: unknown,
): Promise<Response> {
  return fetch(`${conn.addr}/v1/${path}`, {
    method,
    headers: {
      "X-Vault-Token": conn.token,
      ...(body ? { "Content-Type": "application/json" } : {}),
    },
    body: body ? JSON.stringify(body) : undefined,
    cache: "no-store",
  });
}

export type VaultStatus = {
  configured: boolean;
  reachable: boolean;
  addr: string | null;
  mount: string;
  source: "db" | "env" | "none";
  error?: string;
};

/** Ping an explicit Vault connection's health endpoint. */
export async function pingVault(conn: VaultConn): Promise<{ reachable: boolean; error?: string }> {
  try {
    // Validate URL *and* token: lookup-self requires a valid token, so this
    // confirms both reachability and that the token works (unlike sys/health,
    // which is unauthenticated and would pass even with a bad token).
    const res = await vaultFetch(conn, "GET", "auth/token/lookup-self");
    if (res.status === 200) return { reachable: true };
    if (res.status === 403) return { reachable: false, error: "Token rejected (403) — check the Vault token." };
    if (res.status === 503) return { reachable: false, error: "Vault is sealed (503) — unseal it first." };
    if (res.status === 501) return { reachable: false, error: "Vault is not initialised (501)." };
    return { reachable: false, error: `Vault returned ${res.status}.` };
  } catch (e) {
    // Network-level failure — almost always Vault isn't running / not reachable.
    const msg = e instanceof Error ? (e.cause ? `${e.message} (${String(e.cause)})` : e.message) : String(e);
    return {
      reachable: false,
      error: `Could not reach Vault at ${conn.addr} — is it running and reachable from the server? [${msg}]`,
    };
  }
}

/** Connection status for a project's effective Vault (DB row or env). */
export async function vaultStatus(projectId: string | null): Promise<VaultStatus> {
  const conn = await resolveVaultConn(projectId);
  if (!conn) {
    return { configured: false, reachable: false, addr: null, mount: "secret", source: "none" };
  }
  // Callers that need `source` precisely use getVaultConfigView. Report "db"
  // when a projectId was supplied and resolved, else "env".
  const ping = await pingVault(conn);
  return {
    configured: true,
    reachable: ping.reachable,
    addr: conn.addr,
    mount: conn.mount,
    source: projectId ? "db" : "env",
    ...(ping.error ? { error: ping.error } : {}),
  };
}

/** Store an AWS access key + secret for a provider, into its owner's Vault. */
export async function saveAwsKeys(providerId: string, keys: AwsKeys): Promise<void> {
  const conn = await resolveVaultConnForProvider(providerId);
  if (!conn) throw new Error("Vault is not configured. Set up the Vault connection first.");
  const path = `${conn.prefix}/${providerId}`;
  const res = await vaultFetch(conn, "POST", `${conn.mount}/data/${path}`, {
    data: {
      aws_access_key_id: keys.accessKeyId,
      aws_secret_access_key: keys.secretAccessKey,
      ...(keys.region ? { region: keys.region } : {}),
    },
  });
  if (!res.ok) {
    throw new Error(`Vault write failed (${res.status}): ${(await res.text()).slice(0, 300)}`);
  }
}

/** Read an AWS access key + secret back. Returns null if no secret exists. */
export async function getAwsKeys(providerId: string): Promise<AwsKeys | null> {
  const conn = await resolveVaultConnForProvider(providerId);
  if (!conn) return null;
  const path = `${conn.prefix}/${providerId}`;
  const res = await vaultFetch(conn, "GET", `${conn.mount}/data/${path}`);
  if (res.status === 404) return null;
  if (!res.ok) {
    throw new Error(`Vault read failed (${res.status}): ${(await res.text()).slice(0, 300)}`);
  }
  const json = (await res.json()) as { data?: { data?: Record<string, string> } };
  const data = json.data?.data;
  if (!data?.aws_access_key_id || !data?.aws_secret_access_key) return null;
  return {
    accessKeyId: data.aws_access_key_id,
    secretAccessKey: data.aws_secret_access_key,
    region: data.region,
  };
}

/** Remove all versions of a provider's stored secret. Idempotent. */
export async function deleteAwsKeys(providerId: string): Promise<void> {
  const conn = await resolveVaultConnForProvider(providerId);
  if (!conn) return;
  const path = `${conn.prefix}/${providerId}`;
  const res = await vaultFetch(conn, "DELETE", `${conn.mount}/metadata/${path}`);
  if (!res.ok && res.status !== 404) {
    throw new Error(`Vault delete failed (${res.status}): ${(await res.text()).slice(0, 300)}`);
  }
}

File: backend/app/routers/vault.py
import os
import logging
import hvac
from fastapi import APIRouter
from pydantic import BaseModel
from app.core.vault import save_aws_creds_to_vault, get_aws_creds

logger = logging.getLogger(__name__)


class VaultConfig(BaseModel):
    vault_addr: str
    vault_token: str


class AWSCredsPayload(BaseModel):
    aws_access_key_id: str
    aws_secret_access_key: str
    region: str = "us-east-1"


def get_router() -> APIRouter:
    router = APIRouter()

    @router.get("/status")
    def vault_status():
        addr  = os.getenv("VAULT_ADDR", "").strip()
        token = os.getenv("VAULT_TOKEN", "").strip()
        if not addr or not token:
            return {"connected": False, "message": "Vault not configured"}
        try:
            client = hvac.Client(url=addr, token=token)
            if client.is_authenticated():
                return {"connected": True, "addr": addr}
            return {"connected": False, "message": "Invalid or expired token"}
        except Exception as e:
            return {"connected": False, "message": str(e)}

    @router.post("/save")
    def save_vault_config(config: VaultConfig):
        """Save Vault connection config (addr + token) to .env."""
        addr  = config.vault_addr.strip()
        token = config.vault_token.strip()
        if not addr or not token:
            return {"status": "error", "message": "vault_addr and vault_token are required"}
        try:
            client = hvac.Client(url=addr, token=token)
            if not client.is_authenticated():
                return {"status": "error", "message": "Invalid or expired Vault token"}

            os.environ["VAULT_ADDR"]  = addr
            os.environ["VAULT_TOKEN"] = token

            env_path = _resolve_env_path()
            _update_env_file(env_path, "VAULT_ADDR",  addr)
            _update_env_file(env_path, "VAULT_TOKEN", token)

            logger.info(f"✅ Vault config saved (addr={addr})")
            return {"status": "saved", "addr": addr}
        except Exception as e:
            logger.error(f"Failed to save Vault config: {e}", exc_info=True)
            return {"status": "error", "message": str(e)}

    @router.post("/save-aws")
    def save_aws_creds(payload: AWSCredsPayload):
        """
        Store AWS access key + secret into Vault.
        The agent will always fetch from here — never from .env.
        """
        try:
            save_aws_creds_to_vault(
                access_key=payload.aws_access_key_id,
                secret_key=payload.aws_secret_access_key,
                region=payload.region,
            )
            return {"status": "saved", "message": "AWS credentials stored in Vault"}
        except Exception as e:
            logger.error(f"Failed to save AWS creds to Vault: {e}", exc_info=True)
            return {"status": "error", "message": str(e)}

    @router.get("/get-aws")
    def get_aws():
        """
        Verify the agent can read AWS creds from Vault.
        Returns the key ID only (never returns the secret).
        """
        try:
            creds = get_aws_creds()
            return {
                "status": "ok",
                "aws_access_key_id": creds["aws_access_key_id"],
                "region": creds["region"],
                "secret_present": bool(creds["aws_secret_access_key"]),
            }
        except Exception as e:
            return {"status": "error", "message": str(e)}

    return router


# ── Helpers ────────────────────────────────────────────────────────────────────

def _resolve_env_path() -> str:
    explicit = os.getenv("VAULT_ENV_PATH", "")
    if explicit:
        return explicit
    here = os.path.dirname(os.path.abspath(__file__))
    return os.path.join(here, "..", ".env")


def _update_env_file(path: str, key: str, value: str):
    try:
        abs_path = os.path.abspath(path)
        lines = open(abs_path).readlines() if os.path.exists(abs_path) else []
        updated = False
        for i, line in enumerate(lines):
            if line.startswith(f"{key}="):
                lines[i] = f"{key}={value}\n"
                updated = True
                break
        if not updated:
            lines.append(f"{key}={value}\n")
        with open(abs_path, "w") as f:
            f.writelines(lines)
    except Exception as e:
        logger.warning(f"Could not update .env at {path}: {e}")